docs/legacy/developer/api-versioning.md

```diff
@@ -1,464 +1,279 @@
-change on the backend first.  This worked well on the cloud where we
-had control, but fails a lot in the on-prem setting: if we add a new
-end-point to the API, the backend will still be able to handle older
-clients that simply don't know about that end-point, but the client
-won't handle old backends well, since it will try to call the
-end-point, and fail.
+An *API version* is a natural number, represented as `vN`, where `N` is the
+version. For example, version `5` is denoted `v5`.
 
-The problem becomes more complicated still if you think about backends
-talking to other backends in the context of federation: An HTTP server
-will act as a client inside the handler function, and the version it
-responds with may differ from the version it talks to another backend
-it needs to query.
+A backend advertises a set of *supported* API versions, divided into a set of
+*stable* API versions and a set of *development* API versions. These sets can
+be discovered via the `GET /api-version` endpoint, which returns a JSON object
+of the form:
 
-The new approach outlined here introduces API versions to address this
-problem.  Every API version is only compatible with itself, but every
-node in the network can support a *set of API versions*.  A HTTP
-client can query the set of supported versions from an HTTP server,
-and then pick one that works for it.
+```json
+{ "supported": [0, 1, 2, 3, 4],
+  "development": [4],
+  ...
+}
+```
 
-We believe this is a good approach to solve all our API compatibility
-problems both between apps and backends and between backends in the
-context of federation, but we also list some open questions (and
-probably forget to list more).
+where the `...` stands for other fields, which are irrelevant to the purposes
+of API versioning. Development versions (usually one), are listed explicitly.
+Stable versions are all supported versions that are not development.
 
-In the following, we will refer to HTTP clients as "clients", no
-matter whether it is an app or a backend talking to another backend
-(federation); and to HTTP servers as "server", no matter which API is
-it serving (federation or app).
+Stable versions map to well-defined and fixed API *contracts*, which will not
+change over time. That is to say, if two different instances of the backend
+both list version `3` in `supported`, a client can assume that they will accept
+the same exact requests when version `3` is used.
 
+On the other hand, development versions give no such guarantees. When making
+requests using a development version on some backend, the client needs to be
+aware of the corresponding API contract used by that specific backend.
 
-# Versions and servant routing tables
+Of course, development versions are useful while building a new API, but are
+not suitable for production. Backends deployed to production environments
+should disable development versions (and not advertise them in `/api-version`).
```

Comment:
```suggestion
should disable development versions (and not advertise them in `/api-version`).

On the other hand, production clients must disregard all development versions
as unsupported.
```

This would be easier if development versions would not also be listed as versions.  I'm actually in favor of this.  I think it's a lot less error-prone.

Reply:
I went back and forth with this, and in the end I think I prefer the current system, because it places no burden on clients. Clients can just start using V2 now, without any special logic, and as long as we disable it in production, it won't affect anything.

Also, it makes it easier to test, because we only need to make sure that the backend has a development version enabled, and not that the client has turned out development API access as well.

But I can be convinced otherwise. This is not too problematic to change later on, anyway.

Reply:
I have added the following paragraph:
> Similarly, clients that are meant for production use can decide to ignore
development versions on their backend. This is not strictly necessary, but it
can be used as a safeguard against mistakes in deployment.

Reply:
> I went back and forth with this, and in the end I think I prefer the current system, because it places no burden on clients. Clients can just start using V2 now, without any special logic, and as long as we disable it in production, it won't affect anything.
> 
> Also, it makes it easier to test, because we only need to make sure that the backend has a development version enabled, and not that the client has turned out development API access as well.
> 
> But I can be convinced otherwise. This is not too problematic to change later on, anyway.

I'm ok with everything except the last sentence.  :)  So, yes, do it your way.